core/src/main/scala/pureconfig/ConfigConvert.scala

```diff
@@ -155,27 +157,44 @@ object ConfigConvert extends LowPriorityConfigConvertImplicits {
   case class NoValidCoproductChoiceFound(config: ConfigValue)
     extends RuntimeException(s"No valid coproduct type choice found for configuration $config.")
 
-  implicit def cNilConfigConvert: ConfigConvert[CNil] = new ConfigConvert[CNil] {
+  implicit def cNilConfigConvert[Wrapped]: WrappedConfigConvert[Wrapped, CNil] = new WrappedConfigConvert[Wrapped, CNil] {
     override def from(config: ConfigValue): Try[CNil] =
       Failure(NoValidCoproductChoiceFound(config))
 
     override def to(t: CNil): ConfigValue = ConfigFactory.parseMap(Map().asJava).root()
   }
 
-  implicit def coproductConfigConvert[V, T <: Coproduct](
-    implicit vFieldConvert: Lazy[ConfigConvert[V]],
-    tConfigConvert: Lazy[ConfigConvert[T]]): ConfigConvert[V :+: T] =
-    new ConfigConvert[V :+: T] {
+  implicit def coproductConfigConvert[Wrapped, Name <: Symbol, V, T <: Coproduct](
+    implicit coproductHint: CoproductHint[Wrapped],
+    vName: Witness.Aux[Name],
+    vFieldConvert: Lazy[ConfigConvert[V]],
+    tConfigConvert: Lazy[WrappedConfigConvert[Wrapped, T]]): WrappedConfigConvert[Wrapped, FieldType[Name, V] :+: T] =
+    new WrappedConfigConvert[Wrapped, FieldType[Name, V]:+: T] {
+
+      override def from(config: ConfigValue): Try[FieldType[Name, V] :+: T] = config match {
+        case null => Failure(CannotConvertNullException)
+
+        case _ => coproductHint.from(config, vName.value.name) match {
+          case Success(Some(hintConfig)) =>
+            vFieldConvert.value.from(hintConfig) match {
+              case Failure(_) if coproductHint.tryNextOnFail(vName.value.name) =>
+                tConfigConvert.value.from(config).map(s => Inr(s))
 
-      override def from(config: ConfigValue): Try[V :+: T] = {
-        vFieldConvert.value.from(config)
-          .map(s => Inl[V, T](s))
-          .orElse(tConfigConvert.value.from(config).map(s => Inr[V, T](s)))
+              case vTry => vTry.map(v => Inl(field[Name](v)))
+            }
+
+          case Success(None) => tConfigConvert.value.from(config).map(s => Inr(s))
+          case Failure(ex) => Failure(ex)
+        }
       }
 
-      override def to(t: V :+: T): ConfigValue = t match {
-        case Inl(l) => vFieldConvert.value.to(l)
-        case Inr(r) => tConfigConvert.value.to(r)
+      override def to(t: FieldType[Name, V] :+: T): ConfigValue = t match {
+        case Inl(l) =>
+          // Writing a coproduct to a config can fail. Is it worth it to make `to` return a `Try`?
```

Comment:
I still think that `to` should be total, as you. The kind of errors that can make `to` fail are errors that should be able to be checked in compile time or at least when the `ConfigConvert` is instantiated.

The only reason I made this a `Try` is in [CoproductHint.scala#L86](https://github.com/ruippeixotog/pureconfig/blob/cb1435dd4f328973f8cb563e5f18fd29e596b435/core/src/main/scala/pureconfig/CoproductHint.scala#L86). When this happens, _all_ instances of the sealed family fail to serialize, not just part of them. Technically, pureconfig should be able to detect a conflicting key when deriving the `ConfigConvert` instance. It can be improved in a future pull request, I think.

It may be that there are other type of errors I am not thinking of that can cause `to` to fail, but right now we do not have such cases. Conceptually, it seems to me that the function is total - every instance should have a representing config.